File: frontend/src/lib/api/analytics.ts
/** Analytics aggregate fetched by the v1.0.0 dashboard's `/analytics`
 * page. Mirrors the `AnalyticsSummaryOut` Pydantic model on the
 * server (see `bearings.api.models.analytics`). One request renders
 * the whole page; per-card endpoints would multiply round-trips
 * without changing the freshness story. */

import { jsonFetch } from './core';

export type SessionsByDay = {
  /** ISO date 'YYYY-MM-DD'. */
  day: string;
  count: number;
};

export type TopTag = {
  id: number;
  name: string;
  color: string | null;
  session_count: number;
};

export type AnalyticsSummary = {
  total_sessions: number;
  open_sessions: number;
  closed_sessions: number;

  total_messages: number;

  total_input_tokens: number;
  total_output_tokens: number;
  total_cache_read_tokens: number;
  total_cache_creation_tokens: number;
  total_tokens: number;

  total_cost_usd: number;

  /** Always `days` entries long, zero-filled where no sessions
   * were created on a given day. */
  sessions_by_day: SessionsByDay[];

  /** Top 10 tags by session_count desc, then name asc. */
  top_tags: TopTag[];
};

/** Fetch the per-instance analytics rollup. `days` clamps the
 * sessions-by-day time series only; the headline totals are
 * all-time and ignore it. */
export function fetchAnalyticsSummary(
  days = 30,
  fetchImpl: typeof fetch = fetch
): Promise<AnalyticsSummary> {
  return jsonFetch<AnalyticsSummary>(fetchImpl, `/api/analytics/summary?days=${days}`);
}

File: src/bearings/api/models/analytics.py
"""Analytics DTOs for the v1.0.0 dashboard's `/analytics` page.

A single response model bundles every aggregate the page renders so
the frontend can paint the whole surface from one fetch. Per-card
endpoints would multiply round-trips without changing freshness:
the dashboard reads at-rest, refreshing on user action, not as a
live stream.
"""

from __future__ import annotations

from pydantic import BaseModel


class SessionsByDay(BaseModel):
    """One bucket of the sessions-by-day time series. Days with zero
    sessions are zero-filled server-side so the chart axis stays
    continuous (a sparse list reads as 'no data' rather than 'no
    sessions that day')."""

    day: str  # ISO date 'YYYY-MM-DD'
    count: int


class TopTag(BaseModel):
    """One row of the top-N-tags-by-session-count list."""

    id: int
    name: str
    color: str | None = None
    session_count: int


class AnalyticsSummaryOut(BaseModel):
    """Per-instance aggregate snapshot. Counts are non-negative;
    `total_cost_usd` is a float (post-COALESCE, never null)."""

    total_sessions: int
    open_sessions: int
    closed_sessions: int

    total_messages: int

    total_input_tokens: int
    total_output_tokens: int
    total_cache_read_tokens: int
    total_cache_creation_tokens: int
    total_tokens: int

    total_cost_usd: float

    sessions_by_day: list[SessionsByDay]
    top_tags: list[TopTag]
